brainzutils/ratelimit.py

```diff
@@ -0,0 +1,261 @@
+    '''
+
+    limits_timeout = getattr(g, '_' + ratelimit_timeout, 0)
+    if time.time() <= limits_timeout:
+        return
+
+    value = int(cache.get(ratelimit_per_token_key) or '0')
+    if not value:
+        cache.set(ratelimit_per_token_key, ratelimit_per_token_default)
+        value = ratelimit_per_token_default
+    setattr(g, '_' + ratelimit_per_token_key, value)
+
+    value = int(cache.get(ratelimit_per_ip_key) or '0')
+    if not value:
+        cache.set(ratelimit_per_ip_key, ratelimit_per_ip_default)
+        value = ratelimit_per_ip_default
+    setattr(g, '_' + ratelimit_per_ip_key, value)
+
+    value = int(cache.get(ratelimit_window_key) or '0')
+    if not value:
+        cache.set(ratelimit_window_key, ratelimit_window_default)
+        value = ratelimit_window_default
+    setattr(g, '_' + ratelimit_window_key, value)
+
+    setattr(g, '_' + ratelimit_timeout, int(time.time()) + ratelimit_refresh)
+
+
+def get_per_ip_limits():
+    '''
+        Fetch the per IP limits from context/cache
+    '''
+    check_limit_freshness()
+    return {
+            'limit':   getattr(g, '_' + ratelimit_per_ip_key),
+            'window' : getattr(g, '_' + ratelimit_window_key),
+           }
+
+
+def get_per_token_limits():
+    '''
+        Fetch the per token limits from context/cache
+    '''
+    check_limit_freshness()
+    return {
+            'limit':   getattr(g, '_' + ratelimit_per_token_key),
+            'window' : getattr(g, '_' + ratelimit_window_key),
+           }
+
+
+def get_rate_limit_data(request):
+    '''Fetch key for the given request. If an Authorization header is provided,
+       the caller will get a better and personalized rate limit. If no header is provided,
+       the caller will be rate limited by IP, which gets an overall lower rate limit.
+       This should encourage callers to always provide the Authorization token
+    '''
+
+    # If a user verification function is provided, parse the Authorization header and try to look up that user
+    if ratelimit_user_validation:
+        auth_header = request.headers.get('Authorization')
+        if auth_header:
+            auth_token = auth_header[6:]
+            is_valid = ratelimit_user_validation(auth_token)
+            if is_valid:
+                values = get_per_token_limits()
+                values['key'] = auth_token
+                return values
+
+
+    # no valid auth token provided. Look for a remote addr header provided a the proxy
+    # or if that isn't available use the IP address from the header
+    ip = request.environ.get('REMOTE_ADDR', None)
+    if not ip:
+        ip = request.remote_addr
+
+    values = get_per_ip_limits()
+    values['key'] = ip
+    return values
+
+
+def ratelimit():
+    '''
+        This is the decorator that should be applied to all view functions that should be 
+        rate limited.
+    '''
+    def decorator(f):
```

Comment:
For best practises, this should also include `functools.wraps`: https://docs.python.org/3.7/library/functools.html#functools.wraps